File: android/src/main/java/com/margelo/nitro/webrtcplayback/HybridWebRtcPlayback.kt
package com.margelo.nitro.webrtcplayback

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip

/**
 * The JS handle on the playout path. The path itself is installed by
 * [WebRtcPlaybackInitProvider] at process start; this only reads it back and
 * switches the leveller.
 */
@DoNotStrip
@Keep
class HybridWebRtcPlayback : HybridWebRtcPlaybackSpec() {
  override val isInstalled: Boolean
    get() = WebRtcPlaybackInstaller.isInstalled

  override val isPlaying: Boolean
    get() = WebRtcPlaybackInstaller.isPlaying

  override val config: WebRtcPlaybackConfig
    get() {
      val installed = WebRtcPlaybackInstaller.config
      return WebRtcPlaybackConfig(
        playoutDelayMs = installed.playoutDelayMs.toDouble(),
        maxPlayoutDelayMs = installed.maxPlayoutDelayMs.toDouble(),
        leveller = installed.leveller,
        levellerInputGainDb = installed.levellerInputGainDb.toDouble(),
        manageAudioSession = false,
        audioFocus = installed.audioFocus,
        networkResilience = installed.networkResilience,
      )
    }

  override var levellerEnabled: Boolean
    get() = WebRtcPlaybackInstaller.levellerEnabled
    set(value) {
      WebRtcPlaybackInstaller.levellerEnabled = value
    }

  /** DynamicsProcessing exposes no metering, so Android has nothing to report. */
  override fun takeLevels(): PlayoutLevels? = null
}

File: ios/HybridWebRtcPlayback.swift
import NitroModules

/// The JS handle on the playout device. The device itself is installed from
/// `+load` in WRPAudioDevice.m, before any JS runs; this only reads it back and
/// switches the leveller.
class HybridWebRtcPlayback: HybridWebRtcPlaybackSpec {
  var isInstalled: Bool {
    return WRPIsInstalled()
  }

  var isPlaying: Bool {
    return WRPIsPlaying()
  }

  var config: WebRtcPlaybackConfig {
    let installed = WRPGetInstallConfig()
    return WebRtcPlaybackConfig(
      playoutDelayMs: installed.playoutDelayMs,
      maxPlayoutDelayMs: installed.maxPlayoutDelayMs,
      leveller: installed.leveller,
      levellerInputGainDb: installed.levellerInputGainDb,
      manageAudioSession: installed.manageAudioSession,
      audioFocus: false,
      networkResilience: installed.networkResilience
    )
  }

  var levellerEnabled: Bool {
    get { return WRPGetLevellerEnabled() }
    set { WRPSetLevellerEnabled(newValue) }
  }

  func takeLevels() throws -> PlayoutLevels? {
    var input = 0.0
    var output = 0.0
    var reduction = 0.0
    guard WRPTakeLevels(&input, &output, &reduction) else { return nil }
    return PlayoutLevels(inputPeakDb: input, outputPeakDb: output, maxReductionDb: reduction)
  }
}
